contracts/vesting-lp-pcl/src/contract.rs

```diff
@@ -0,0 +1,170 @@
+    msg: InstantiateMsg,
+) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let owner = deps.api.addr_validate(&msg.owner)?;
+    CONFIG.save(
+        deps.storage,
+        &Config {
+            owner,
+            vesting_token: Option::from(msg.vesting_token),
+            token_info_manager: deps.api.addr_validate(&msg.token_info_manager)?,
+
+            extensions: Extensions {
+                historical: true,
+                managed: true,
+                with_managers: true,
+            },
+        },
+    )?;
+    for m in msg.vesting_managers.iter() {
+        let ma = deps.api.addr_validate(m)?;
+        VESTING_MANAGERS.save(deps.storage, ma, &())?;
+    }
+    XYK_VESTING_LP_CONTRACT.save(
+        deps.storage,
+        &deps
+            .api
+            .addr_validate(&msg.xyk_vesting_lp_contract.clone())?,
+    )?;
+    Ok(Response::default())
+}
+
+/// Exposes execute functions available in the contract.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(msg) => receive_cw20(deps, env, info, msg),
+        _ => {
+            let base_msg: BaseExecute = msg.into();
+            base_execute(deps, env, info, base_msg)
+        }
+    }
+}
+
+/// Receives a message of type [`Cw20HookMsg`] and processes it depending on the received template.
+///
+/// * **cw20_msg** CW20 message to process.
+fn receive_cw20(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    cw20_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let vesting_token = get_vesting_token(&config)?;
+
+    // Permission check
+    if !is_sender_whitelisted(
+        deps.storage,
+        &config,
+        &deps.api.addr_validate(&cw20_msg.sender)?,
+    ) || token_asset_info(info.sender) != vesting_token
+    {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    match from_binary(&cw20_msg.msg)? {
+        Cw20HookMsg::RegisterVestingAccounts { vesting_accounts } => {
+            register_vesting_accounts(deps, vesting_accounts, cw20_msg.amount, env.block.height)
+        }
+        Cw20HookMsg::MigrateXYKLiquidity {
+            user_address_raw,
+            user_vesting_info,
+        } => handle_migrate_xyk_liquidity(deps, env, user_address_raw, user_vesting_info),
+    }
+}
+
+fn is_sender_whitelisted(store: &mut dyn Storage, config: &Config, sender: &Addr) -> bool {
+    if *sender == config.owner {
+        return true;
+    }
+    let xyk_vesting_lp_contract = XYK_VESTING_LP_CONTRACT.load(store).unwrap();
+    if *sender == xyk_vesting_lp_contract {
+        return true;
+    }
+    if VESTING_MANAGERS.has(store, sender.clone()) {
+        return true;
+    }
+
+    false
+}
+
+fn handle_migrate_xyk_liquidity(
```

Comment:
Shouldn't we check that received amount of tokens equals to amount of tokens in vesting schedules?

Reply:
<img width="713" alt="Снимок экрана 2024-02-20 в 10 35 38" src="https://github.com/neutron-org/neutron-tge-contracts/assets/26491894/2a384f13-5b9d-4aa0-b6e4-3047d23fc66c">
<img width="784" alt="Снимок экрана 2024-02-20 в 10 35 25" src="https://github.com/neutron-org/neutron-tge-contracts/assets/26491894/6021f974-f681-4ae4-af21-6f4c4324060b">

This check will be unnecessary because the values of end point amount and amount to send are both set from a single variable.